data/srv/epages/eproot/Perl/man/man3/accessors.3
.\" Automatically generated by Pod::Man 2.23 (Pod::Simple 3.14)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.ie \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.el \{\
.    de IX
..
.\}
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "accessors 3"
.TH accessors 3 "2014-08-08" "perl v5.12.3" "User Contributed Perl Documentation"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
accessors \- create accessor methods in caller's package.
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
.Vb 2
\&  package Foo;
\&  use accessors qw( foo bar baz );
\&
\&  my $obj = bless {}, \*(AqFoo\*(Aq;
\&
\&  # generates chaining accessors
\&  # that you can set like this:
\&  $obj\->foo( \*(Aqhello \*(Aq )
\&      \->bar( \*(Aqworld\*(Aq )
\&      \->baz( "!\en" );
\&
\&  # you get the values by passing no params:
\&  print $obj\->foo, $obj\->bar, $obj\->baz;
.Ve
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
The \fBaccessors\fR pragma lets you create simple accessors at compile-time.
.PP
This saves you from writing them by hand, which tends to result in
\&\fIcut-n-paste\fR errors and a mess of duplicated code.  It can also help you
reduce the ammount of unwanted \fIdirect-variable access\fR that may creep into
your codebase when you're feeling lazy.  \fBaccessors\fR was designed with
laziness in mind.
.PP
Method-chaining accessors are generated by default.  Note that you can still
use accessors::chained directly for reasons of backwards compatability.
.PP
See accessors::classic for accessors that always return the current value if
you don't like method chaining.
.SH "GENERATED METHODS"
.IX Header "GENERATED METHODS"
\&\fBaccessors\fR will generate methods that return the current object on set:
.PP
.Vb 5
\&  sub foo {
\&      my $self = shift;
\&      if (@_) { $self\->{\-foo} = shift; return $self; }
\&      else    { return $self\->{\-foo}; }
\&  }
.Ve
.PP
This way they can be \fIchained\fR together.
.SS "Why prepend the dash?"
.IX Subsection "Why prepend the dash?"
The dash (\f(CW\*(C`\-\*(C'\fR) is prepended to the property name for a few reasons:
.IP "\(bu" 4
interoperability with Error.
.IP "\(bu" 4
to make it difficult to accidentally access the property directly ala:
.Sp
.Vb 3
\&  use accessors qw( foo );
\&  $obj\->{foo};  # prevents this by mistake
\&  $obj\->foo;    # when you probably meant this
.Ve
.Sp
(this might sound woolly, but it's easy enough to do).
.IP "\(bu" 4
syntactic sugar (this \fIis\fR woolly :).
.PP
You shouldn't care too much about how the property is stored anyway \- if you do,
you're likely trying to do something special (and should really consider writing
the accessors out long hand), or it's simply a matter of preference in which
case you can use accessors::classic, or sub-class this module.
.SH "PERFORMANCE"
.IX Header "PERFORMANCE"
There is \fBlittle-to-no performace hit\fR when using generated accessors; in
fact there is \fBusually a performance gain\fR.
.IP "\(bu" 4
typically \fI10\-30% faster\fR than hard-coded accessors (like the above example).
.IP "\(bu" 4
typically \fI1\-15% slower\fR than \fIoptimized\fR accessors (less readable).
.IP "\(bu" 4
typically a \fIsmall\fR performance hit at startup (accessors are created at
compile-time).
.IP "\(bu" 4
uses the same anonymous sub to reduce memory consumption (sometimes by 80%).
.PP
See the benchmark tests included with this distribution for more details.
.SH "MOTIVATION"
.IX Header "MOTIVATION"
The main difference between the \fBaccessors\fR pragma and other accessor
generators is \fBsimplicity\fR.
.IP "\(bu" 4
interface
.Sp
\&\fBuse accessors qw( ... )\fR is as easy as it gets.
.IP "\(bu" 4
a pragma
.Sp
it fits in nicely with the \fBbase\fR pragma:
.Sp
.Vb 2
\&  use base      qw( Some::Class );
\&  use accessors qw( foo bar baz );
.Ve
.Sp
and accessors get created at compile-time.
.IP "\(bu" 4
no bells and whistles
.Sp
The module is extensible instead.
.SH "SUB-CLASSING"
.IX Header "SUB-CLASSING"
If you prefer a different style of accessor or you need to do something more
complicated, there's nothing to stop you from sub-classing.  It should be
pretty easy.  Look through accessors::classic, accessors::ro, and
accessors::rw to see how it's done.
.SH "CAVEATS"
.IX Header "CAVEATS"
Classes using blessed scalarrefs, arrayrefs, etc. are not supported for sake
of simplicity.  Only hashrefs are supported.
.SH "THANKS"
.IX Header "THANKS"
Thanks to Michael G. Schwern for indirectly inspiring this module, and for his
feedback & suggestions.
.PP
Also to Paul Makepeace and David Wright for showing me faster accessors, to
chocolateboy for his contributions, the \s-1CPAN\s0 Testers for their bug reports,
and to James Duncan and people on London.pm for their feedback.
.SH "AUTHOR"
.IX Header "AUTHOR"
Steve Purkis <spurkis@cpan.org>
.SH "SEE ALSO"
.IX Header "SEE ALSO"
accessors::classic, accessors::chained
.PP
Similar and related modules:
.PP
base,
fields,
Class::Accessor,
Class::Struct,
Class::Methodmaker,
Class::Generate,
Class::Class,
Class::Tangram,
Object::Tiny


data/srv/epages/eproot/Perl/lib/site_perl/accessors.pm
=head1 NAME

accessors - create accessor methods in caller's package.

=head1 SYNOPSIS

  package Foo;
  use accessors qw( foo bar baz );

  my $obj = bless {}, 'Foo';

  # generates chaining accessors
  # that you can set like this:
  $obj->foo( 'hello ' )
      ->bar( 'world' )
      ->baz( "!\n" );

  # you get the values by passing no params:
  print $obj->foo, $obj->bar, $obj->baz;

=cut

package accessors;

use 5.006;
use strict;
use warnings::register;

our $VERSION  = '1.01';
our $REVISION = (split(/ /, ' $Revision: 1.22 $ '))[2];

our $Debug        = 0;
our $ExportLevel  = 0;
our @InvalidNames = qw( BEGIN CHECK INIT END DESTROY AUTOLOAD );

use constant style => 'chained';

sub import {
    my $class   = shift;
    my $callpkg = caller( $class->ExportLevel );

    my @properties = @_ or return;

    $class->create_accessors_for( $callpkg, @properties );
}

sub create_accessors_for {
    my $class   = shift;
    my $callpkg = shift;

    warn( 'creating ' . $class->style . ' accessors( ',
          join(' ',@_)," ) in pkg '$callpkg'" ) if $class->Debug;

    foreach my $property (@_) {
        my $accessor = "$callpkg\::$property";
        die( "can't create $accessor - '$property' is not a valid name!" )
          unless $class->isa_valid_name( $property );
        warn( "creating " . $class->style . " accessor: $accessor\n" ) if
          $class->Debug > 1;
        $class->create_accessor( $accessor, $property );
    }

    return $class;
}

sub create_accessor {
    my ($class, $accessor, $property) = @_;
    $property = "-$property";
    # set/get is slightly faster if we eval instead of using a closure + anon
    # sub, but the difference is marginal (~5%), and this uses less memory...
    no strict 'refs';
    *{$accessor} = sub {
        (@_ > 1)
          ? ($_[0]->{$property} = $_[1], return $_[0])
          : $_[0]->{$property};
    };
}

sub isa_valid_name {
    my ($class, $property) = @_;
    return unless $property =~ /^(?!\d)\w+$/;
    return if grep {$property eq $_} $class->InvalidNames;
    return 1;
}

##
## on the off-chance that someone will sub-class:
##

## don't like studly caps for sub-names, but stick with Exporter-like style...
sub Debug        { $Debug; }
sub ExportLevel  { $ExportLevel }
sub InvalidNames { @InvalidNames }

1;

__END__

=head1 DESCRIPTION

The B<accessors> pragma lets you create simple accessors at compile-time.

This saves you from writing them by hand, which tends to result in
I<cut-n-paste> errors and a mess of duplicated code.  It can also help you
reduce the ammount of unwanted I<direct-variable access> that may creep into
your codebase when you're feeling lazy.  B<accessors> was designed with
laziness in mind.

Method-chaining accessors are generated by default.  Note that you can still
use L<accessors::chained> directly for reasons of backwards compatability.

See L<accessors::classic> for accessors that always return the current value if
you don't like method chaining.

=head1 GENERATED METHODS

B<accessors> will generate methods that return the current object on set:

  sub foo {
      my $self = shift;
      if (@_) { $self->{-foo} = shift; return $self; }
      else    { return $self->{-foo}; }
  }

This way they can be I<chained> together.

=head2 Why prepend the dash?

The dash (C<->) is prepended to the property name for a few reasons:

=over 4

=item *

interoperability with L<Error>.

=item *

to make it difficult to accidentally access the property directly ala:

  use accessors qw( foo );
  $obj->{foo};  # prevents this by mistake
  $obj->foo;    # when you probably meant this

(this might sound woolly, but it's easy enough to do).

=item *

syntactic sugar (this I<is> woolly :).

=back

You shouldn't care too much about how the property is stored anyway - if you do,
you're likely trying to do something special (and should really consider writing
the accessors out long hand), or it's simply a matter of preference in which
case you can use L<accessors::classic>, or sub-class this module.

=head1 PERFORMANCE

There is B<little-to-no performace hit> when using generated accessors; in
fact there is B<usually a performance gain>.

=over 4

=item *

typically I<10-30% faster> than hard-coded accessors (like the above example).

=item *

typically I<1-15% slower> than I<optimized> accessors (less readable).

=item *

typically a I<small> performance hit at startup (accessors are created at
compile-time).

=item *

uses the same anonymous sub to reduce memory consumption (sometimes by 80%).

=back

See the benchmark tests included with this distribution for more details.

=head1 MOTIVATION

The main difference between the B<accessors> pragma and other accessor
generators is B<simplicity>.

=over 4

=item * interface

B<use accessors qw( ... )> is as easy as it gets.

=item * a pragma

it fits in nicely with the B<base> pragma:

  use base      qw( Some::Class );
  use accessors qw( foo bar baz );

and accessors get created at compile-time.

=item * no bells and whistles

The module is extensible instead.

=back

=head1 SUB-CLASSING

If you prefer a different style of accessor or you need to do something more
complicated, there's nothing to stop you from sub-classing.  It should be
pretty easy.  Look through L<accessors::classic>, L<accessors::ro>, and
L<accessors::rw> to see how it's done.

=head1 CAVEATS

Classes using blessed scalarrefs, arrayrefs, etc. are not supported for sake
of simplicity.  Only hashrefs are supported.

=head1 THANKS

Thanks to Michael G. Schwern for indirectly inspiring this module, and for his
feedback & suggestions.

Also to Paul Makepeace and David Wright for showing me faster accessors, to
chocolateboy for his contributions, the CPAN Testers for their bug reports,
and to James Duncan and people on London.pm for their feedback.

=head1 AUTHOR

Steve Purkis <spurkis@cpan.org>

=head1 SEE ALSO

L<accessors::classic>, L<accessors::chained>

Similar and related modules:

L<base>,
L<fields>,
L<Class::Accessor>,
L<Class::Struct>,
L<Class::Methodmaker>,
L<Class::Generate>,
L<Class::Class>,
L<Class::Tangram>,
L<Object::Tiny>

=cut
